src/nb_type.cpp

```diff
@@ -336,101 +335,234 @@ static void inst_dealloc(PyObject *self) {
+        void *foreign_bindings = t ? nb_load_acquire(t->foreign_bindings)
+                                   : nullptr;
+        *has_foreign = t && foreign_bindings &&
+                       (nb_is_seq(foreign_bindings) ||
+                        ((pymb_binding *) foreign_bindings)->framework !=
+                            internals_->foreign_self);
+    }
+#else
+    (void) has_foreign;
+#endif
+
+    return t;
 }
 
-void nb_type_unregister(type_data *t) noexcept {
+static bool nb_type_update_cache(nb_type_map_fast &cache,
+                                 const std::type_info *type,
+                                 nb_alias_seq *more_types,
+                                 void *value) {
+    bool found = cache.update(type, value);
+    while (more_types) {
+        found |= cache.update(more_types->value, value);
+        more_types = more_types->next;
+    }
+    return found;
+}
+
+void nb_type_update_c2p_fast(const std::type_info *type, void *value) noexcept {
+    // internals must be locked by the caller
+    nb_internals *internals_ = internals;
+    auto it_alias = internals_->types_in_c2p_fast.find(type);
+    if (it_alias != internals_->types_in_c2p_fast.end()) {
+#if defined(NB_FREE_THREADED)
+        for (nb_type_map_per_thread *cache =
+                 internals_->type_c2p_per_thread_head;
+             cache; cache = cache->next) {
+            nb_type_update_cache(*cache->lock(), it_alias->first,
+                                 (nb_alias_seq *) it_alias->second,
+                                 value);
+        }
+        // We can't require that we found a match, because the type might
+        // have been cached only by a thread that has since exited.
+#else
+        bool found = nb_type_update_cache(
+            internals_->type_c2p_fast, it_alias->first,
+            (nb_alias_seq *) it_alias->second, value);
+        check(found, "nanobind::detail::nb_type_update_c2p_fast(\"%s\"): "
+                     "types_in_c2p_fast and type_c2p_fast are inconsistent",
+              type_name(type));
+#endif
+    }
+}
+
+bool nb_type_register(type_data *t, type_data **conflict) noexcept {
     nb_internals *internals_ = internals;
     nb_type_map_slow &type_c2p_slow = internals_->type_c2p_slow;
 
     lock_internals guard(internals_);
-    size_t n_del_slow = type_c2p_slow.erase(t->type);
-
-#if defined(NB_FREE_THREADED)
-    // In free-threaded mode, stale type information remains in the
-    // 'type_c2p_fast' TLS. This data structure is eventually deallocated
-    // when the thread terminates.
-    //
-    // In principle, this is dangerous because the user could delete a type
-    // binding from a module at runtime, causing the associated
-    // Python type object to be freed. If a function then attempts to return
-    // a value with such a de-registered type, nanobind should raise an
-    // exception, which requires knowing that the entry in 'type_c2p_fast'
-    // has become invalid in the meantime.
-    //
```

Comment:
Given the removal of this paragraph, does it mean that the new version is more robust wrt. (potentially non-immortal) type objects in free-threaded Python extensions in the future?